File: Indexed Search/Basic Indexed Search/Sparse Index Search/SparseIndexSearch.py
import math


def sparse_index_search(
        arr,
        target):

    n = len(arr)

    step = int(
        math.sqrt(n)
    )

    index = []

    for i in range(
            0,
            n,
            step):

        index.append(
            (
                arr[i],
                i
            )
        )

    start = 0

    for key, pos in index:

        if key > target:
            break

        start = pos

    end = min(
        start
        +
        step
        -
        1,
        n - 1
    )

    for i in range(
            start,
            end + 1):

        if arr[i] == target:
            return i

    return -1


arr = [
2,5,8,12,
16,20,24,
28,32,40
]

print(
sparse_index_search(
arr,
24
)
)

File: Indexed Search/Basic Indexed Search/Sparse Index Search/SparseIndexSearch.cpp
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

struct Sparse {

int key;

int pos;

};

int sparseSearch(
vector<int>& arr,
int target
){

int n=
arr.size();

int step=
sqrt(n);

vector<Sparse>
index;

for(
int i=0;
i<n;
i+=step
){

index.push_back(
{
arr[i],
i
}
);

}

int start=0;

for(
auto&
x:
index
){

if(
x.key
>
target
)
break;

start=
x.pos;

}

int end=
min(
start
+
step
-
1,
n-1
);

for(
int i=start;
i<=end;
i++
){

if(
arr[i]
==
target
)
return i;

}

return -1;

}

int main(){

vector<int>
arr=
{
2,5,8,12,
16,20,24,
28,32,40
};

cout
<<
sparseSearch(
arr,
24
);

}
